### PR TITLE
breaking(layout_column_wrap): Make `width` an optional argument

### PR DESCRIPTION
**Pull request by gadenbuie**
Avoids items overflowing the parent layout container. With this rule, `width` becomes the point at which a new column is created, i.e. if `width = "400px"` then new columns are added to a row if there is at least 400px of space.

**Breaking:** This PR also makes `width` an optional argument and moves it to after the `...`. A [code search shows that most people name the `width` argument](https://github.com/search?q=language:r+layout_column_wrap&type=code), but in case they haven't we'll test to see if the first argument to `layout_column_wrap()` looks like a valid `width` value (a single numeric value or a valid CSS unit) and we'll emit a deprecation warning and use that first value as `width`.

Note: in the deprecation warning I'm assuming this breaking change will cause our next version to be v0.6.0 (in any case, we should review our deprecation warnings before the next release to make sure we have the right version in them).

Some code to try out this feature:

```r
x <- lapply(1:3, \(x) card(lorem::ipsum(x)))

layout_column_wrap(!!!x)

# Explicit, named `width`
layout_column_wrap(width = 1/2, !!!x)
layout_column_wrap(width = "300px", !!!x)
layout_column_wrap(width = 400, !!!x)

# Implied as the first argument,
# leads to deprecation warning but gives same result
layout_column_wrap(1/2, !!!x)
layout_column_wrap("300px", !!!x)
layout_column_wrap(400, !!!x)
```

### CODE REVIEW

File: R/layout.R [hunk 1]
```diff
@@ -110,6 +138,16 @@ layout_column_wrap <- function(
   )
 }
 
+is_probably_a_css_unit <- function(x) {
+  if (!length(x) == 1) return(FALSE)
```
Comment (cpsievert) [inline R/layout.R:142]:
```suggestion
  if (length(x) != 1) return(FALSE)
```

Comment (gadenbuie) [reply R/layout.R:142]: 🤦‍♂️

File: R/layout.R [hunk 2]
```diff
@@ -31,18 +37,23 @@
...
 #'
+#' # To use larger column widths by default, set `width`.
+#' # This example has 3 columns when the screen is at least 900px wide:
+#' layout_column_wrap(width = "300px", x, x, x)
+#'
+#' @export
 layout_column_wrap <- function(
-  width,
   ...,
+  width = "200px",
```
Comment (cpsievert) [inline R/layout.R:56]: Seems like we have room to explore better default behavior. I haven't fully thought this through, but what if a value of `NA` yielded something like:

```css
grid-template-columns: repeat(auto-fit, minmax(clamp(0px, 200px, 100%), 1fr))
```

This way, if you dump a bunch of elements that are less than 200px wide, you don't get a bunch of empty space (e.g., `layout_column_wrap(!!!replicate(10, actionButton("foo", "Button"), simplify = F))`)

Looking at this also makes me wonder if we should've had `align-content: center;` on the grid from day one 🤔

Comment (gadenbuie) [reply R/layout.R:56]: My gut reaction is that `actionButton()` (or shiny inputs in general) should be full width when used as direct children of `layout_column_wrap()`. I recognize that that sort of side-steps the point of your example, but I do think that in many other cases we wouldn't want the clamp behavior.

More broadly, I think we want uniformly even columns from `layout_column_wrap()`, at least when using its public arguments. (Meaning: if you want a non-even column layout you'd define the `grid-template-columns` property via inline styles.) But I think the current proposed default is easier to reason about and closer to what most people want (i.e. put these things in columns but don't let them be smaller than X pixels wide, unless there isn't enough room of course).

**Comment by cpsievert** [comment]
I'm on board with these changes, but I'd like to see this PR split into 2: one for the `min(%s, 100%)` change (with a clear examples/explanation of where it makes a difference, as well as a dedicated bullet for it in the NEWS) and then another PR for the breaking change.

Also, feel free to use your own examples, but here is one that's fairly simple, but demonstrates the fix:

```r
# Open me on a medium sized screen (i.e., between about 800-500px) 
# and notice how the cards "bleed outside" of their containing column

my_card <- card(lorem::ipsum(2), max_height = 100)

page_fixed(
  layout_column_wrap(
    width = "800px", 
    class = "border border-danger",
    !!!replicate(3, my_card, simplify = F)
  )
)
```

**Comment by gadenbuie** [comment]
Closing this PR in favor of a smaller, more targeted PR.

**Pull request closed without merging**